### ISSUE TITLE
Flaky specs in `style/class_check_spec.rb` and `version_spec.rb`

### ISSUE THREAD

**Issue by Earlopain** [rubocop/rubocop#13870]
Ref https://github.com/rubocop/rubocop/pull/13869#issuecomment-2668864645, running on https://github.com/rubocop/rubocop/commit/f91dda2e4ff7bbf6f2e6bdaaff4ad231663dab74:

```sh
$ bundle exec rspec ./spec/rubocop/cop/style/class_check_spec.rb[1:1:1,1:1:2,1:2:1,1:2:2] ./spec/rubocop/version_spec.rb[1:2:4:1,1:2:5:1] --seed 8365
Run options: include {ids: {"./spec/rubocop/cop/style/class_check_spec.rb" => ["1:1:1", "1:1:2", "1:2:1", "1:2:2"], "./spec/rubocop/version_spec.rb" => ["1:2:4:1", "1:2:5:1"]}}

Randomized with seed 8365
..FFFF

Failures:

  1) RuboCop::Cop::Style::ClassCheck when enforced style is kind_of? registers an offense for `is_a?` and corrects to `kind_of?`
     Failure/Error: raise "Unknown style #{s} selected!" unless supported_styles.include?(s)
     
     RuntimeError:
       Unknown style kind_of? selected!
     # ./lib/rubocop/cop/mixin/configurable_enforced_style.rb:82:in 'RuboCop::Cop::ConfigurableEnforcedStyle#style'
     # ./lib/rubocop/cop/style/class_check.rb:34:in 'RuboCop::Cop::Style::ClassCheck#on_send'
     # ./lib/rubocop/cop/commissioner.rb:143:in 'Kernel#public_send'
     # ./lib/rubocop/cop/commissioner.rb:143:in 'block (2 levels) in RuboCop::Cop::Commissioner#trigger_restricted_cops'
     # ./lib/rubocop/cop/commissioner.rb:171:in 'RuboCop::Cop::Commissioner#with_cop_error_handling'
     # ./lib/rubocop/cop/commissioner.rb:142:in 'block in RuboCop::Cop::Commissioner#trigger_restricted_cops'
     # ./lib/rubocop/cop/commissioner.rb:141:in 'Array#each'
     # ./lib/rubocop/cop/commissioner.rb:141:in 'RuboCop::Cop::Commissioner#trigger_restricted_cops'
     # ./lib/rubocop/cop/commissioner.rb:70:in 'RuboCop::Cop::Commissioner#on_send'
     # /home/earlopain/Documents/rubocop-ast/lib/rubocop/ast/traversal.rb:20:in 'RuboCop::AST::Traversal#walk'
     # ./lib/rubocop/cop/commissioner.rb:87:in 'RuboCop::Cop::Commissioner#investigate'
     # ./lib/rubocop/cop/team.rb:174:in 'RuboCop::Cop::Team#investigate_partial'
     # ./lib/rubocop/cop/team.rb:101:in 'RuboCop::Cop::Team#investigate'
     # ./lib/rubocop/rspec/cop_helper.rb:88:in 'CopHelper#_investigate'
     # ./lib/rubocop/rspec/expect_offense.rb:121:in 'RuboCop::RSpec::ExpectOffense#expect_offense'
     # ./spec/rubocop/cop/style/class_check_spec.rb:34:in 'block (3 levels) in <top (required)>'

  2) RuboCop::Cop::Style::ClassCheck when enforced style is kind_of? registers an offense for `is_a?` is safe navigation called and corrects to `kind_of?`
     Failure/Error: raise "Unknown style #{s} selected!" unless supported_styles.include?(s)
     
     RuntimeError:
       Unknown style kind_of? selected!
     # ./lib/rubocop/cop/mixin/configurable_enforced_style.rb:82:in 'RuboCop::Cop::ConfigurableEnforcedStyle#style'
     # ./lib/rubocop/cop/style/class_check.rb:34:in 'RuboCop::Cop::Style::ClassCheck#on_send'
     # ./lib/rubocop/cop/commissioner.rb:143:in 'Kernel#public_send'
     # ./lib/rubocop/cop/commissioner.rb:143:in 'block (2 levels) in RuboCop::Cop::Commissioner#trigger_restricted_cops'
     # ./lib/rubocop/cop/commissioner.rb:171:in 'RuboCop::Cop::Commissioner#with_cop_error_handling'
     # ./lib/rubocop/cop/commissioner.rb:142:in 'block in RuboCop::Cop::Commissioner#trigger_restricted_cops'
     # ./lib/rubocop/cop/commissioner.rb:141:in 'Array#each'
     # ./lib/rubocop/cop/commissioner.rb:141:in 'RuboCop::Cop::Commissioner#trigger_restricted_cops'
     # ./lib/rubocop/cop/commissioner.rb:70:in 'RuboCop::Cop::Commissioner#on_csend'
     # /home/earlopain/Documents/rubocop-ast/lib/rubocop/ast/traversal.rb:20:in 'RuboCop::AST::Traversal#walk'
     # ./lib/rubocop/cop/commissioner.rb:87:in 'RuboCop::Cop::Commissioner#investigate'
     # ./lib/rubocop/cop/team.rb:174:in 'RuboCop::Cop::Team#investigate_partial'
     # ./lib/rubocop/cop/team.rb:101:in 'RuboCop::Cop::Team#investigate'
     # ./lib/rubocop/rspec/cop_helper.rb:88:in 'CopHelper#_investigate'
     # ./lib/rubocop/rspec/expect_offense.rb:121:in 'RuboCop::RSpec::ExpectOffense#expect_offense'
     # ./spec/rubocop/cop/style/class_check_spec.rb:45:in 'block (3 levels) in <top (required)>'

  3) RuboCop::Cop::Style::ClassCheck when enforced style is `is_a?` registers an offense for `kind_of?` is safe navigation called and corrects to `is_a?`
     Failure/Error: raise "Unknown style #{s} selected!" unless supported_styles.include?(s)
     
     RuntimeError:
       Unknown style is_a? selected!
     # ./lib/rubocop/cop/mixin/configurable_enforced_style.rb:82:in 'RuboCop::Cop::ConfigurableEnforcedStyle#style'
     # ./lib/rubocop/cop/style/class_check.rb:34:in 'RuboCop::Cop::Style::ClassCheck#on_send'
     # ./lib/rubocop/cop/commissioner.rb:143:in 'Kernel#public_send'
     # ./lib/rubocop/cop/commissioner.rb:143:in 'block (2 levels) in RuboCop::Cop::Commissioner#trigger_restricted_cops'
     # ./lib/rubocop/cop/commissioner.rb:171:in 'RuboCop::Cop::Commissioner#with_cop_error_handling'
     # ./lib/rubocop/cop/commissioner.rb:142:in 'block in RuboCop::Cop::Commissioner#trigger_restricted_cops'
     # ./lib/rubocop/cop/commissioner.rb:141:in 'Array#each'
     # ./lib/rubocop/cop/commissioner.rb:141:in 'RuboCop::Cop::Commissioner#trigger_restricted_cops'
     # ./lib/rubocop/cop/commissioner.rb:70:in 'RuboCop::Cop::Commissioner#on_csend'
     # /home/earlopain/Documents/rubocop-ast/lib/rubocop/ast/traversal.rb:20:in 'RuboCop::AST::Traversal#walk'
     # ./lib/rubocop/cop/commissioner.rb:87:in 'RuboCop::Cop::Commissioner#investigate'
     # ./lib/rubocop/cop/team.rb:174:in 'RuboCop::Cop::Team#investigate_partial'
     # ./lib/rubocop/cop/team.rb:101:in 'RuboCop::Cop::Team#investigate'
     # ./lib/rubocop/rspec/cop_helper.rb:88:in 'CopHelper#_investigate'
     # ./lib/rubocop/rspec/expect_offense.rb:121:in 'RuboCop::RSpec::ExpectOffense#expect_offense'
     # ./spec/rubocop/cop/style/class_check_spec.rb:19:in 'block (3 levels) in <top (required)>'

  4) RuboCop::Cop::Style::ClassCheck when enforced style is `is_a?` registers an offense for `kind_of?` and corrects to `is_a?`
     Failure/Error: raise "Unknown style #{s} selected!" unless supported_styles.include?(s)
     
     RuntimeError:
       Unknown style is_a? selected!
     # ./lib/rubocop/cop/mixin/configurable_enforced_style.rb:82:in 'RuboCop::Cop::ConfigurableEnforcedStyle#style'
     # ./lib/rubocop/cop/style/class_check.rb:34:in 'RuboCop::Cop::Style::ClassCheck#on_send'
     # ./lib/rubocop/cop/commissioner.rb:143:in 'Kernel#public_send'
     # ./lib/rubocop/cop/commissioner.rb:143:in 'block (2 levels) in RuboCop::Cop::Commissioner#trigger_restricted_cops'
     # ./lib/rubocop/cop/commissioner.rb:171:in 'RuboCop::Cop::Commissioner#with_cop_error_handling'
     # ./lib/rubocop/cop/commissioner.rb:142:in 'block in RuboCop::Cop::Commissioner#trigger_restricted_cops'
     # ./lib/rubocop/cop/commissioner.rb:141:in 'Array#each'
     # ./lib/rubocop/cop/commissioner.rb:141:in 'RuboCop::Cop::Commissioner#trigger_restricted_cops'
     # ./lib/rubocop/cop/commissioner.rb:70:in 'RuboCop::Cop::Commissioner#on_send'
     # /home/earlopain/Documents/rubocop-ast/lib/rubocop/ast/traversal.rb:20:in 'RuboCop::AST::Traversal#walk'
     # ./lib/rubocop/cop/commissioner.rb:87:in 'RuboCop::Cop::Commissioner#investigate'
     # ./lib/rubocop/cop/team.rb:174:in 'RuboCop::Cop::Team#investigate_partial'
     # ./lib/rubocop/cop/team.rb:101:in 'RuboCop::Cop::Team#investigate'
     # ./lib/rubocop/rspec/cop_helper.rb:88:in 'CopHelper#_investigate'
     # ./lib/rubocop/rspec/expect_offense.rb:121:in 'RuboCop::RSpec::ExpectOffense#expect_offense'
     # ./spec/rubocop/cop/style/class_check_spec.rb:8:in 'block (3 levels) in <top (required)>'

Finished in 0.31446 seconds (files took 1.12 seconds to load)
6 examples, 4 failures

Failed examples:

rspec ./spec/rubocop/cop/style/class_check_spec.rb:33 # RuboCop::Cop::Style::ClassCheck when enforced style is kind_of? registers an offense for `is_a?` and corrects to `kind_of?`
rspec ./spec/rubocop/cop/style/class_check_spec.rb:44 # RuboCop::Cop::Style::ClassCheck when enforced style is kind_of? registers an offense for `is_a?` is safe navigation called and corrects to `kind_of?`
rspec ./spec/rubocop/cop/style/class_check_spec.rb:18 # RuboCop::Cop::Style::ClassCheck when enforced style is `is_a?` registers an offense for `kind_of?` is safe navigation called and corrects to `is_a?`
rspec ./spec/rubocop/cop/style/class_check_spec.rb:7 # RuboCop::Cop::Style::ClassCheck when enforced style is `is_a?` registers an offense for `kind_of?` and corrects to `is_a?`

Randomized with seed 8365
```

**Comment by dvandersluis** [comment]
Got this here too on the JRuby run: https://github.com/rubocop/rubocop/actions/runs/13416752592/job/37479335428?pr=13871

**Comment by koic** [comment]
Since it can be reproduced with the seed 8365, this does not seem to be an environment-specific issue.

**Comment by dvandersluis** [comment]
It happened in my example above with seed 17329 FYI.

**Comment by dvandersluis** [comment]
OK I think I figured it out.

When `version_spec` loads the plugins specified in its tests, it eventually calls `Plugin::ConfigurationIntegrator#combine_rubocop_configs`, which **overrides** `ConfigLoader.default_configuration` with the merged configuration. Because the plugin adds configuration, that gets stored as the new default configuration, and is never undone.

Secondly, as @Earlopain discovered, the configuration for `Style/ClassCheck` gets overridden by the configuration for `RSpec/ClassCheck`. I'm not 100% sure why this happening, because it doesn't happen when running rubocop normally. It seems like the plugin cops are not being required in the test before the plugin config is merged, which makes `Registry#qualified_cop_name` for `RSpec/ClassCheck` return `Style/ClassCheck` (because that cop **is** already registered, and the RSpec one isn't, it gets returned as if the wrong namespace is specified). 

Thus, the configuration for `RSpec/ClassCheck` gets merged when processing the plugin over top of that of `Style/ClassCheck`. And then, since `ConfigLoader.default_configuration` is permanently overridden, it persists and causes the tests for `Style/ClassCheck` to fail if run after the `version_spec` tests because the configuration has changed.

PR is coming to fix the tests, but it'd probably be good to revisit overriding default configuration -- we don't seem to have to do that for `require`.
